### DISCUSSION TITLE
Customizing "Go Home"

### DISCUSSION THREAD

**Discussion by ykazakov** [coder/code-server#2245] [Ideas] [answered]
Is it possible to customise the "Go Home" link in the application menu?

![Screenshot 2020-10-28 at 12 19 51](https://user-images.githubusercontent.com/2140361/97430736-99ca9780-1919-11eb-98f2-6752a6734451.png)

Currently it leads to the github page of vscode. I would like to instead forward to my web application from which the editor was opened.

**Comment by code-asher** [comment]
It appears to be hardcoded. I was going to remove it but we could
definitely make it customizable instead.

**Reply by ykazakov** [reply]
How would you customize it? As command-line options or in some application settings?

Looking for something similar/related noticed that on the welcome screen (Menu>Help>Welcome) many links below `Help` are non-functional: 
- Printable keyboard cheatsheet
- Introductory videos
- Tips and tricks
- Product documentation

Are they hardcoded as well or can they be configured?
I just think that adding lots of command-line options is not the best solution to customize something like this.

**Reply by code-asher** [reply]
Good catch, all those welcome links are supposed to work so I think we'd
categorize them as bugs and get them fixed.

A `--home` command-line flag could work, and that would also make it
possible to specify `home` in ~/.config/code-server/config.yaml.

**Reply by bruwozniak** [reply]
I was looking for exactly such option, so definitely a +1 here

**Comment by ykazakov** [comment]
Maybe the links did not suppose to work because they are specific to vscode?
The `--home` flag seems a reasonable option, I was just wondering how (other) links are set in vscode and to what extent are they customisable.

**Reply by code-asher** [reply]
Fair question. I'm not completely sure so we'll need to poke around and see. If we do end up needing to customize additional links maybe we can use our config file instead of command line flags.

**Comment by bruwozniak** [comment]
@code-asher any prospect of moving forward on this? I am happy to contribute a PR but I could use some pointers on where to look for the values

**Reply by code-asher** [reply]
I don't think it'll be on my radar for a while yet. A PR would definitely be welcome!

In the patch there is an `initialize` method in `lib/vscode/vs/server/node/server.ts` that returns an object with a `workbenchWebConfiguration` property. That fulfills an interface which is a partial copy of VS Code's `IWorkbenchConstructionOptions`. In `IWorkbenchConstructionOptions` there is a `homeIndicator` property we could copy to our interface then add to `workbenchWebConfiguration`. 

Unfortunately VS Code then overrides that with a hardcoded home indicator in `lib/vscode/src/vs/code/browser/workbench/workbench.ts#L458` so we'll also need to remove that override.

Assuming that works, from there we'd just need to add `home` to our CLI options and related interfaces.

**Comment by jsjoeio** [comment, marked as answer]
This has been added and should be available in the next release! 

See [the upcoming docs](https://github.com/cdr/code-server/blob/master/doc/FAQ.md#how-do-i-customize-the-go-home-button) to learn how you'll be able to utilize it.

Thanks @ykazakov for starting this discussion!

PR #2435

**Reply by jsjoeio** [reply]
Update: this will be removed when https://github.com/cdr/code-server/pull/2862 is merged (it was removed upstream)

**Reply by ykazakov** [reply]
Would it be now possible to similarly customize the logout link? Currently this leads to `/login`. We use an external authenticator (by setting the `key` cookie and the corresponding value of `HASHED_PASSWORD`) so the code-server login page should not be used.

<img width="179" alt="Screenshot 2021-04-09 at 09 34 51" src="https://user-images.githubusercontent.com/2140361/114145702-fabdb900-9916-11eb-8149-114c9bfb6f80.png">

**Reply by code-asher** [reply]
How do you handle authentication when a user first visits code-server? Some kind of redirect if the cookie isn't set or something? I'm curious if you could do a similar thing for `/login` (if that's what you're doing).

**Reply by ykazakov** [reply]
We use code-server in [JupyterHub](https://jupyter.org/hub) with a [jupyter-server-proxy plugin](https://github.com/jupyterhub/jupyter-server-proxy) and it almost runs out of the box. The plugin is responsible for starting code-server on some free port and proxying it to path like myhost/user/$username/code-server/ (once the user with $username is authenticated with JupyterHub). When the proxy is started, we generate a random cookie and pass it to the sever using the `HASHED_PASSWORD` environment variable. So if the user visits this link, the cookie is *always* set (using the `Set-Cookie` header) and one never gets to see the code-server login page.

The problem with the `Log out` menu is that it links to `myhost/login` not the *proxied path* `myhost/user/$username/code-server/login` (which would then immediately redirect to `myhost/user/$username/code-server/code-server` since the cookie is always set). I think, the jupyter-server-proxy plugin does not rewrite links. As far as I see, `/login` is the only (non-base) path used by code-server.

**Reply by code-asher** [reply]
Aaahh, that makes sense. That reminds me that we need to make the logout redirect relative. That should sort this use case, I think?

I think we'll end up fixing this as part of https://github.com/cdr/code-server/issues/2984#issuecomment-812652886

**Reply by ykazakov** [reply]
OK, that should at least fix the broken link.

However, it will probably not handle logout in our case. As I mentioned, the cookie is always set by the web server (tornado used by JupyterHub), even if it gets deleted by code-server. I can possibly add some handler to the web server for the `/login` path, but I cannot *distinguish* based on this path whether the user wants to *login* or *logout*.

A bit better solution, from my opinion, would be to add another path, say `/logout` which handles cookie delete and other cleanup, and redirects to `/login`. Then I can handle this path from the web server and redirect to whatever I want.

Of course, if the logout path could be configured then it would be even easier to handle for us. But I can see that adding another command line option might be not the cleanest solution.

**Reply by bruwozniak** [reply]
going back to the original topic of `Go Home` menu item, is there any chance for this to make a comeback at some point? We're running code-server in JupyterHub and it was very useful to be able to go back to the main hub menu.

**Reply by code-asher** [reply]
@ykazakov That's perfect! I wanted us to redirect to a `/logout` page so we could do the cleanup on the server instead of via JavaScript. We'll do that as part of #2984.

@bruwozniak I think at the moment we don't have any plans for that, we mostly only make patches to VS Code for bug fixes. But we did plan on having some kind of overlay at some point, maybe that would be a good place to add a home link.

Or...actually, we soon plan on shipping a built-in extension with VS Code to provide some helpful functionality much like the remote extensions do. So we could probably add a menu item through the extension that way we don't need to mess with VS Code itself.

**Reply by bruwozniak** [reply]
That sounds valid, I suppose it would be quite easy to use the extension API to add such menu item. Looking forward to the extension code release then 👍

**Reply by ykazakov** [reply]
Excellent plan! I guess for now I can handle `/logout`. Look forward to the extension as well!

**Comment by aiweaver** [comment]
What is the progress of this issue? This is a feature I need!

**Reply by code-asher** [reply]
The "Go Home" link seems to be gone now, so I do not think we will be patching anything here.

I think it is possible to add menu items via an extension, so that might be the best way to implement, if someone wants to try.
